### ISSUE TITLE
Which version is best ?

### ISSUE THREAD

**Issue by dotNetAthlete** [gothinkster/aspnetcore-realworld-example-app#35]
I'm building a new asp.net web api using .net core 2.1.
To begin with I used NuGet packages of 2.1.1 and I ran into dependency issues.
I now seem to have a stable project that uses 2.1.0-preview1-final.

The NuGet gallery website allows different versions to be used but which version is best ?

What is the best practice surrounding this ?

**Comment by adamhathcock** [comment]
The bigger question is: what are you using that’s dependent on the preview bits?

**Comment by adamhathcock** [comment]
You should be using 2.1.1

**Comment by dotNetAthlete** [comment]
I agree, but last time I used 2.1.1, I ran into a dependency issue with a NuGet package that needed to target 2.1.0, so what is the 'best practice' when faced with such an issue ?

**Comment by dotNetAthlete** [comment]
Ok, so if I delete the current NuGet packages and install the 2.1.1 current versions, will I risk breaking anything in my project ???

**Comment by dotNetAthlete** [comment]
The answer is YES, so thanks very much :(

PM> Install-Package Microsoft.AspNetCore.App -Version 2.1.1
Restoring packages for C:\Users\everettr\source\repos\OCM_API\OCM_API\OCM_API.csproj...
NU1608: Detected package version outside of dependency constraint: Microsoft.CodeAnalysis.CSharp.Workspaces 2.6.1 requires Microsoft.CodeAnalysis.CSharp (= 2.6.1) but version Microsoft.CodeAnalysis.CSharp 2.8.0 was resolved.
Install-Package : NU1107: Version conflict detected for Microsoft.CodeAnalysis.Common. Reference the package directly from the project to resolve this issue. 
 OCM_API -> Microsoft.AspNetCore.App 2.1.1 -> Microsoft.AspNetCore.Mvc.Razor 2.1.1 -> Microsoft.CodeAnalysis.CSharp 2.8.0 -> Microsoft.CodeAnalysis.Common (= 2.8.0) 
 OCM_API -> Microsoft.VisualStudio.Web.CodeGeneration.Design 2.1.0-preview1-final -> Microsoft.VisualStudio.Web.CodeGenerators.Mvc 2.1.0-preview1-final -> 
Microsoft.VisualStudio.Web.CodeGeneration 2.1.0-preview1-final -> Microsoft.VisualStudio.Web.CodeGeneration.EntityFrameworkCore 2.1.0-preview1-final -> 
Microsoft.VisualStudio.Web.CodeGeneration.Core 2.1.0-preview1-final -> Microsoft.VisualStudio.Web.CodeGeneration.Templating 2.1.0-preview1-final -> 
Microsoft.VisualStudio.Web.CodeGeneration.Utils 2.1.0-preview1-final -> Microsoft.CodeAnalysis.CSharp.Workspaces 2.6.1 -> Microsoft.CodeAnalysis.Workspaces.Common 2.6.1 -> 
Microsoft.CodeAnalysis.Common (= 2.6.1).
At line:1 char:1
+ Install-Package Microsoft.AspNetCore.App -Version 2.1.1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
 
Install-Package : Package restore failed. Rolling back package changes for 'OCM_API'.
At line:1 char:1
+ Install-Package Microsoft.AspNetCore.App -Version 2.1.1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
 
Time Elapsed: 00:00:02.1836939
PM>

**Comment by adamhathcock** [comment]
Looks like you're directly referencing `Microsoft.VisualStudio.Web.CodeGeneration.Design 2.1.0-preview1-final ` which should be 2.1.1 as it exists here: https://www.nuget.org/packages/Microsoft.VisualStudio.Web.CodeGeneration.Design/

**Comment by dotNetAthlete** [comment]
In the end I had to completely re create my project from scratch as it
wouldn’t target the correct ask.

Seems like github and nuget has created a new DLL hell


We need a node package version manager

**Comment by adamhathcock** [comment]
NPM isn't much better.  You can and do still run into transitive dependency issues when referencing an old package (which was your case).   csproj and Nuget aren't as simple but they're not terrible either.  I encourage you to do a diff on your new csproj vs the one that didn't work to understand what the issue was.
